File: HW1/HIS_Specialization.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
#read the image into matrix
img = cv2.imread('./image/1.jpg')
img_2 = cv2.imread('./image/2.jpg')
img_2 = cv2.cvtColor(img_2,cv2.COLOR_RGB2GRAY)
img = cv2.cvtColor(img,cv2.COLOR_RGB2GRAY)

hist = np.zeros([256,1])
hist_2 = np.zeros([256,1])

for i in range (256):
    hist[i] = np.sum(img==i)
    hist_2[i] = np.sum(img_2==i)
#normalization
hist = hist/sum(hist)
hist_2 = hist_2/sum(hist_2)
#calculate the cumulative histogram
Cu_hist = np.cumsum(hist)
Cu_hist_2 = np.cumsum(hist_2)

#build the look up table
I = np.zeros([256,2])
for i in range(256):
    I[i] = np.min(np.argwhere(Cu_hist_2-Cu_hist[i]==np.min(abs(Cu_hist_2-Cu_hist[i]))))[0][0]
#plot plot plot!

plt.figure(1,figsize=(9,9))
plt.subplot(321)
plt.plot(hist)
plt.title('Histogram of first image')

plt.subplot(322)
plt.plot(Cu_hist)
plt.title('Cumulative Histogram of first image')

plt.subplot(323)
plt.plot(hist_2)
plt.title('Histogram of second image')

plt.subplot(324)
plt.plot(Cu_hist_2)
plt.title('Cumulative Histogram of second image')

plt.subplot(325)
plt.plt(I)
plt.title('The look up table I')

plt.show()

File: HW1/His_specialization.m
%read image and convert into grey
img1 = imread('./image./1.jpg');
img2 = imread('./image./2.jpg');
img1 = rgb2gray(img1);
img2 = rgb2gray(img2);
im1 = img1;
im2 = img2;
%make histogram
hist1 = zeros(1,256);
hist2 = zeros(1,256);
for i = 1:256
    hist1(i) = sum(sum(img1==i-1));
    hist2(i) = sum(sum(img2==i-1));
end
hist1 = hist1/sum(hist1);
hist2 = hist2/sum(hist2);
cu_hist1 = cumsum(hist1);
cu_hist2 = cumsum(hist2);
I = zeros(1,256);
%build look up table
for i = 1:256
    I(i) = min(find(abs(cu_hist2-cu_hist1(i))==min(abs(cu_hist2-cu_hist1(i)))));
    %map img1
    img1(find(im1==i-1)) = I(i);
end

%calculate histogram of changed image
hist3 = zeros(1,256);
for k = 1:256
    hist3(k) = sum(sum(img1==k-1));
end
hist3 = hist3/sum(hist3);


subplot(331);
plot(hist1);
subplot(332);
plot(cu_hist1);

subplot(333);
imshow(im1);

subplot(334);
plot(hist2);

subplot(335);
plot(cu_hist2);

subplot(336);
imshow(im2);

subplot(337);
plot(hist3);

subplot(338);
plot(I);

subplot(339);
imshow(img1);




